Adjust quarter date range selection options to work with Visitor stats

ThomazFB opened the pull request:
Summary
==========
Fix issue #7750 by introducing a proper way to request the `Last quarter` and `quarter to date` date range selections for the `Visitors and Views` data. This is done by doing a dedicated request to Quarter-related range selections.

How to Test
==========
1. Go to your store jetpack stats (URL should be /wp-admin/admin.php?page=stats)
2. Verify the Month views and visitors data chart
3. Go to the Analytics Hub and select `Last Quarter` or `Quarter to Date` date range options
4. Check if the data displayed in the Visitors and Views card in the Analytics Hub corresponds to the data sum of the respective three months of the quarter displayed in your store stats.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ThomazFB committed:
Make request fetching amount limit a controlled parameter

ThomazFB committed:
add constants for fetching limit

ThomazFB committed:
Force recent visitors fetching to only capture one data element

ThomazFB committed:
Distinct quarter date range selections

ThomazFB committed:
Refactor timePeriod acquisition

ThomazFB committed:
Configure quarter selection flag for visitors updating function

ThomazFB committed:
Simplify quarter date range event distinction selection

ThomazFB committed:
Simplify quarter date range event distinction selection even further

ThomazFB committed:
Switch between quarter and usual date requests

ThomazFB committed:
Introduce quarter fetching structure

ThomazFB committed:
Define an extension function to acquire the day before a given date

ThomazFB committed:
Refactor fetchQuarterVisitorData function

ThomazFB committed:
Fix lint issues

ThomazFB committed:
Fix unit tests

ThomazFB commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -203,6 +210,33 @@ class AnalyticsRepository @Inject constructor(
+        selectedRange: AnalyticTimePeriod,
+        fetchStrategy: FetchStrategy
+    ): VisitorsResult {
+        val startDate = dateRange.getSelectedPeriod().from.theDayBeforeIt()
```
We need to move the start date one day before so the month comparisons don't match. E.g:

Quarter months response:
2022-10-01
2022-11-01
2022-12-01

start date:
2022-10-01

Since we use a date comparison through the `before` function in the filter below, we want the interval to be closed on the start date, hence the `theDayBeforeIt` operation to ensure that, making the start date to become 2022-09-30.

ThomazFB committed:
Refactor the quarter selection value acquisition

ThomazFB committed:
Improve AnalyticsViewModel imports

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -203,6 +210,33 @@ class AnalyticsRepository @Inject constructor(
+        fetchStrategy: FetchStrategy
+    ): VisitorsResult {
+        val startDate = dateRange.getSelectedPeriod().from.theDayBeforeIt()
+        val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
```
This is nothing that needs to be changed, I just want to mention that on FluxC, we have a DateUtils.getDateFromString() function that receives a String and returns a Date. It will use Locale.ROOT instead of Locale.getDefault(), though.

ThomazFB replied:
Oh, that's true! Great suggestion! Adjusted here: https://github.com/woocommerce/woocommerce-android/pull/7776/commits/e1cc4c3ad0a781516f39b2d6c4336047ed260787

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsRepository.kt:
```diff
@@ -203,6 +210,33 @@ class AnalyticsRepository @Inject constructor(
+                        views.toInt()
+                    )
+                )
+            } ?: VisitorsResult.VisitorsError
```
I believe that with the changes proposed [here](https://wooengage.wordpress.com/2022/11/09/analytics-hub-jetpack-stats-request-for-visitor-and-views/):  we can avoid this filter and unify the request. We just would need to select the correct `unit` and `quantity`, and it will even work for custom date ranges. Something we can work on on a different PR so we can ship this to our merchants.

ThomazFB replied:
I agree. This is a disposable code. We should get rid of it once we have your proposed solution.

atorresveiga reviewed (approved):
Great work @ThomazFB!
The code is looking good and working as expected. I just leave a tiny comment about parsing a Date from String but feel free dismiss it if our intention is to use the `Locale.getDefault()`

LGTM :shipit:

atorresveiga commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/analytics/AnalyticsViewModel.kt:
```diff
@@ -281,6 +292,10 @@ class AnalyticsViewModel @Inject constructor(
                 return@launch
             }
 
+            val isQuarterSelection = state.value.analyticsDateRangeSelectorState.selectedPeriod
+                .let { AnalyticTimePeriod.from(it) }
+                .let { (it == QUARTER_TO_DATE) || (it == LAST_QUARTER) }
+
```
Is there a reason why we are not using the `timePeriod` variable below?
I believe we can change it to:
`val isQuarterSelection = timePeriod == QUARTER_TO_DATE || timePeriod == LAST_QUARTER`
And have the same result. WDYT?

ThomazFB replied:
I feel dumb for not noticing we're acquiring the very same datePeriod right below that code 😅. Thanks for catching it! Fixed here: https://github.com/woocommerce/woocommerce-android/pull/7776/commits/5491a16c5ab2caa87861fea27b9857aed515b15f

ThomazFB committed:
Use dateUtils for string to date conversion inside AnalyticsRepository

ThomazFB committed:
Refactor how isQuarterSelection is evaluated

ThomazFB committed:
Fix lint issues

ThomazFB commented:
@atorresveiga I've addressed all your suggestions. Thanks for this fast response! Let me know if you have anything else you think we should change! Thanks 🙇
